crates/core/src/host/wasm_common/module_host_actor.rs

```diff
@@ -589,47 +579,71 @@ impl<T: WasmInstance> WasmModuleInstance<T> {
-                }
+                };
+
+                (status, metrics)
             }
             Ok(Err(errmsg)) => {
-                stdb.rollback_mut_tx(&ctx, tx);
+                let metrics = stdb.rollback_mut_tx(&ctx, tx);
 
                 log::info!("reducer returned error: {errmsg}");
 
-                EventStatus::Failed(errmsg.into())
+                let status = EventStatus::Failed(errmsg.into());
+                (status, metrics)
             }
             Ok(Ok(())) => {
-                if let Some(tx_data) = stdb.commit_tx(&ctx, tx).unwrap() {
-                    EventStatus::Committed(DatabaseUpdate::from_writes(&tx_data))
-                } else {
-                    todo!("Write skew, you need to implement retries my man, T-dawg.");
-                }
+                let (tx_data, metrics) = stdb.commit_tx(&ctx, tx).unwrap();
+                let status = EventStatus::Committed(DatabaseUpdate::from_writes(&tx_data));
+                (status, metrics)
             }
         };
 
         let outcome = ReducerOutcome::from(&status);
 
-        let event = ModuleEvent {
+        let event = Arc::new(ModuleEvent {
             timestamp,
             caller_identity,
             caller_address: caller_address_opt,
-            function_call: ModuleFunctionCall {
-                reducer: reducer_name.to_owned(),
-                args,
-            },
+            function_call,
             status,
             energy_quanta_used: energy.used,
             host_execution_duration: timings.total_duration,
             request_id,
             timer,
-        };
-        self.info
-            .subscriptions
-            .blocking_broadcast_event(client.as_deref(), &subscriptions, Arc::new(event));
+        });
+        let ctx_incr_update = ExecutionContext::incremental_update(address);
+        let _metrics_incr_update = self.info.subscriptions.blocking_broadcast_event(
+            &ctx_incr_update,
+            client.as_deref(),
+            &subscriptions,
+            &event,
+        );
+        drop(subscriptions);
+
+        let reducer_span = reducer_span.entered();
+        reducer_span
+            .record("timings.total_duration", tracing::field::debug(timings.total_duration))
+            .record("energy.used", tracing::field::debug(energy.used));
+
+        const FRAME_LEN_60FPS: Duration = const_unwrap(Duration::from_secs(1).checked_div(60));
+        if timings.total_duration > FRAME_LEN_60FPS {
+            // If we can't get your reducer done in a single frame we should debug it.
+            tracing::debug!(
+                message = "Long running reducer finished executing",
+                reducer_name,
+                ?timings.total_duration,
+            );
+        }
+        reducer_span.exit();
+
+        WORKER_METRICS
+            .reducer_compute_time
+            .with_label_values(&address, reducer_name)
+            .observe(timings.total_duration.as_secs_f64());
```

Comment:
Any specific reason this is not in the `defer` above? Looks like only data from `ExecuteResult` is used.